[C++ Frontend] Remove size() from BatchDataset and templatize IndexType

goldsborough opened the pull request:
This PR brings to changes to the recently landed C++ Frontend dataloader:

1. Removes the `size()` method from `BatchDataset`. This makes it cleaner to implement unsized ("infinite stream") datasets. The method was not used much beyond initial configuration.
2. Makes the index type of a dataset a template parameter of `BatchDataset` and `Sampler`. This essentially allows custom index types instead of only `vector<size_t>`. This greatly improves flexibility.

See the `InfiniteStreamDataset` and `TestIndex` datasets in the tests for what this enables.

Some additional minor updates and code movements too.

@apaszke @SsnL

goldsborough committed:
Remove size() from BatchDataset and templatize IndexType

SsnL commented on torch/csrc/api/include/torch/data/dataloader.h:
```diff
@@ -37,29 +37,20 @@ class DataLoader {
       : options_(std::move(options)),
```
Maybe I missed it, but is there a way to get the sampler from a data loader? One thing people often want to do, and we don't really support properly in Python API, is to dynamically update the sampler.

goldsborough replied:
I don't think that would work too well because `Sampler` is a template parameter, so it's type is rather hard to change any way

SsnL replied:
Oh by dynamic I mean to change the state of the sampler. For instance, adjust an internal state vector of the sampler.

goldsborough replied:
I can provide access to it yes

SsnL commented on torch/csrc/api/include/torch/data/samplers/random.h:
```diff
@@ -14,48 +10,24 @@ namespace data {
-    index_ += index_batch.size();
-    return index_batch;
-  }
+  optional<std::vector<size_t>> next(size_t batch_size) override;
```
dumb question and probably never relevant in practice: does optional support nesting, i.e., `optional<optional<my_index_t>>`?

goldsborough replied:
Yep it works fine

apaszke commented:
I didn't have time to review this yet, but I'm concerned that our C++ dataset API starts diverging from the Python one significantly. We should stop this and bring them together, or we should stop calling this a C++ interface to PyTorch.

goldsborough commented:
@apaszke that sounds a bit dramatic :) Let me give some more context on these two changes. @SsnL, @soumith  and I actually sat together before meeting with some partners to discuss changes to the C++ dataloader (which MSFT wants to use), and how we can reconcile them with the Python dataloader. We want both to have the same features/functionality and @SsnL is working on making the `size()` optional in Python too.

1. We want to support the use case of infinite stream datasets cleanly. Sure, you can always have `size()` be `INT_MAX`, but (1) this is not a clean API and (2) if you ever accidentally used the (default) `RandomSampler`, you'd have a bit of a memory problem. By making the concept of unsized datasets more "first class", we can verify, for example, that you don't try to create a `RandomSampler` from an unsized dataset.

2. This change actually makes the C++ frontend dataloader more like Python, because you can now have indices other than `size_t`. In Python this all works through duck typing of course. So in C++ we allow this now too, which yields much greater flexibility.

I hope this clarifies things and let me know if you have more thoughts.

goldsborough committed:
Fix CanUseCustomTypeAsIndexType

goldsborough commented:
Requesting review please: @apaszke @ezyang

apaszke commented on torch/csrc/api/include/torch/data/datasets/map.h:
```diff
@@ -16,21 +16,24 @@ namespace datasets {
   /// Gets a batch from the source dataset and applies the transform to it,
   /// returning the result.
-  OutputBatchType get_batch(ArrayRef<size_t> indices) override {
+  OutputBatchType get_batch(IndexType indices) override {
```
You're calling this type `Index` everywhere else. Would be good to keep it consistent (again -- it shows up in error messages).

apaszke commented on torch/csrc/api/include/torch/data/samplers/base.h:
```diff
@@ -15,37 +15,29 @@ class InputArchive;
   virtual void reset() = 0;
 
-  /// Returns the next batch of indices if possible, or an empty optional if the
+  /// Returns the next index if possible, or an empty optional if the
```
I really wish this was called `BatchIndex`

apaszke commented on torch/csrc/api/include/torch/data/samplers/base.h:
```diff
@@ -15,37 +15,29 @@ class InputArchive;
+template <typename Index = std::vector<size_t>>
 class Sampler {
  public:
+  using IndexType = Index;
```
Why can't you simply use `Index` here?

goldsborough replied:
These typedefs are to allow fetching the type later through `typename Sampler::IndexType`

apaszke replied:
Sure I understand their purpose. The problem is that you sometimes call those type aliases `X` and sometimes `XType` which is quite inconsistent.

apaszke commented on torch/csrc/api/include/torch/data/samplers/custom_index.h:
```diff
@@ -0,0 +1,17 @@
+  virtual ~CustomIndex() = default;
+
+  /// The number of elements accessed by this index.
+  virtual size_t size() const = 0;
```
To be honest this is a very confusing API. Index is usually something that refers to a single entity, not a number of them. Again, having a `Batch` part somewhere would clarify that it's some kind of a list-like thing.

apaszke commented on torch/csrc/api/include/torch/data/samplers/sequential.h:
```diff
@@ -1,67 +1,43 @@
-  size_t index() const noexcept {
-    return index_;
-  }
+  size_t index() const noexcept;
```
I assume that both `RandomSampler` and `IndexSampler` methods were simply moved to `.cpp` files verbatim, is that correct?

apaszke commented on torch/csrc/api/include/torch/data/samplers/stream.h:
```diff
@@ -0,0 +1,62 @@
+  size_t size() const noexcept override;
+  operator size_t() const noexcept;
+  size_t size_;
+};
```
I don't think this API is very well suited towards streaming datasets. Basically, I would think of such a dataset as an (potentially infinite) list that's read-once. Once you pop an element, it's gone forever. However, if there's a deterministic process that generates that stream, I'd imagine that no matter how I get to the `n`-th index, the sample that's returned will always be the same. If I understand correctly, this implementation simply ignores the indices completely, and takes whatever first `batch_size` elements at each step are available?

That would be fine if a user implemented this, but it's not fine in this case I think. It would be much better to represent a `CustomIndex` for this dataset as `base` offset + `batch_size` of elements that should be taken from the stream (and the stream asserts that at next iter `base` will point to `base + batch_size` from this one).

goldsborough replied:
> deterministic process

is a pretty hard constraint here, no? This `StreamDataset` should work with something like `/etc/urandom` -- no determinism at all, infinite, and all we do is ask for `batch_size` elements. The `index` kept in the `StreamDataset` is purely to keep track of progress to know how many elements were sampled. I will rename it to `progress` in fact. Do you see what I mean? What do you think?

apaszke replied:
Saying that the nice properties in the deterministic case don't hold in the non-deterministic case (and are equivalent to this), and concluding that we're going to go with something that doesn't ever work is weird tbh.

apaszke replied:
Ok. What you're really describing is that what we call an index (which is some kind of a unique identifier for a dataset element; a key if you will), can really be used as some kind of a metadata that describes what kind of a batch should be returned from the dataset. That's fair, and I don't want to block this PR any further, but we might still want to think about the naming, because it's not really an index.

apaszke reviewed (changes requested):
Haven't checked the tests. I think the API for streams feels a bit like a hack and could be improved to make it much nicer. There are also some rough edges around making the dataset size optional.

If we're going to merge this change, I'd really like us to consider how would we integrate those things into our Python API. Those two interfaces already have different semantics, and I don't want them to diverge any further.

goldsborough commented on torch/csrc/api/include/torch/data/dataloader.h:
```diff
@@ -225,13 +214,13 @@ class DataLoader {
       return nullopt;
     }
-    AT_ASSERT(!indices->empty());
+    AT_ASSERT(indices->size() > 0);
```
Otherwise I can't do this line: `indices->size() < options_.batch_size && options_.drop_last`

goldsborough commented on torch/csrc/api/include/torch/data/dataloader.h:
```diff
@@ -298,9 +287,13 @@ template <
+  AT_CHECK(
+      size.has_value(),
+      "Expected the dataset to be sized in "
+      "order to construct the Sampler");
```
The function has a SFINAE guard to only be called when the sampler is constructible from `size_t`, which is assumed to mean it is constructible from the size of the dataset. I should maybe add a separate base class for this instead, in case `size_t` is used for something else

goldsborough commented on torch/csrc/api/include/torch/data/datasets/base.h:
```diff
@@ -88,6 +85,9 @@ class Dataset : public BatchDataset<Self, std::vector<SingleExample>> {
 };
+
+template <typename Self, typename Batch = std::vector<Example<>>>
+using StreamDataset = BatchDataset<Self, Batch, size_t>;
```
The index type to a stream is the batch size (`size_t`) for the number of samples to fetch from the infinite stream. That's all this typedef sets

goldsborough replied:
I will document this better

goldsborough committed:
Rename Index -> BatchIndex

goldsborough committed:
Rename index -> examples_retrieved_so_far in StreamSampler

goldsborough commented:
@apaszke I've renamed `Index` to `BatchIndex` and documented `StreamSampler`. I added a comment about why `StreamSampler` was written in the "just give me `batch_size` values" way. Let me know what you think.

goldsborough commented:
@apaszke what do you think about the StreamSampler issue? I don't want to have this PR linger too long, there are people at microsoft waiting for it

goldsborough committed:
IndexBatch -> BatchRequest
